src/natcap/invest/ndr/ndr_core.pyx

```diff
@@ -60,211 +48,46 @@ def ndr_eff_calculation(
-                        intermediate_retention = (
-                            (neighbor_effective_retention * current_step_factor) +
-                            (retention_eff_lulc * (1 - current_step_factor)))
-
-                    # Case 3: the other 2 cases have not been hit.
-                    else:
-                        intermediate_retention = neighbor_effective_retention
-
-                    working_retention_eff += (
-                        intermediate_retention * neighbor.flow_proportion)
-
-                if has_outflow:
-                    working_retention_eff /= flow_dir_sum
-                    effective_retention_raster.set(
-                        global_col, global_row, working_retention_eff)
-                else:
-                    raise Exception("got to a cell that has no outflow!")
-            # search upslope to see if we need to push a cell on the stack
-            # for i in range(8):
-            up_neighbors = UpslopeNeighbors(
-                Pixel(mfd_flow_direction_raster, global_col, global_row))
-            for neighbor in up_neighbors:
-                neighbor_outflow_dir = INFLOW_OFFSETS[neighbor.direction]
-                neighbor_outflow_dir_mask = 1 << neighbor_outflow_dir
-                neighbor_process_flow_dir = <int>(
-                    to_process_flow_directions_raster.get(
-                        neighbor.x, neighbor.y))
-                if neighbor_process_flow_dir == 0:
-                    # skip, due to loop invariant this must be a nodata pixel
-                    continue
-                if neighbor_process_flow_dir & neighbor_outflow_dir_mask == 0:
-                    # no outflow
-                    continue
-                # mask out the outflow dir that this iteration processed
-                neighbor_process_flow_dir &= ~neighbor_outflow_dir_mask
-                to_process_flow_directions_raster.set(
-                    neighbor.x, neighbor.y, neighbor_process_flow_dir)
-                if neighbor_process_flow_dir == 0:
-                    # if 0 then all downslope have been processed,
-                    # push on stack, otherwise another downslope pixel will
-                    # pick it up
-                    processing_stack.push(neighbor.y * n_cols + neighbor.x)
-
-    stream_raster.close()
-    crit_len_raster.close()
-    retention_eff_lulc_raster.close()
-    effective_retention_raster.close()
-    mfd_flow_direction_raster.close()
-    to_process_flow_directions_raster.close()
+    args = [
+        flow_direction_path.encode('utf-8'),
+        stream_path.encode('utf-8'),
+        retention_eff_lulc_path.encode('utf-8'),
+        crit_len_path.encode('utf-8'),
+        to_process_flow_directions_path.encode('utf-8'),
+        effective_retention_path.encode('utf-8')]
+
+    if algorithm == 'MFD':
+        run_effective_retention[MFD](*args)
+    else: # D8
+        run_effective_retention[D8](*args)
```

Comment:
I get it now!  That is really pretty neat.  So the implication is that the compiler is, effectively, rendering MFD and D8 versions of the underlying functions to allow us to call either one at runtime.  Awesome.